[FLINK-38770][runtime] Fix parallelism overrides ignored in Application Mode

barakbaron12 opened the pull request:
## What is the purpose of the change

This PR fixes a bug where `pipeline.jobvertex-parallelism-overrides` configuration was completely ignored when running jobs in Application Mode with Flink 2.0+. This broke the Flink Kubernetes Operator's autoscaler functionality.

## Brief change log

- Created `ParallelismOverrideUtil` to centralize override application logic
- Modified all scheduler factories (DefaultScheduler, AdaptiveScheduler, AdaptiveBatchScheduler) to apply overrides after StreamGraph→JobGraph conversion
- Removed duplicate override logic from Dispatcher.java
- Ensured correct configuration precedence: job config > job master config

## Verifying this change

This change added tests and can be verified as follows:

- Added 7 unit tests in `ParallelismOverrideUtilTest` covering all override scenarios
- Added 4 integration tests in `ParallelismOverridesITCase` verifying end-to-end behavior across different scheduler types
- Modified existing integration tests in `AdaptiveBatchSchedulerITCase` and `AdaptiveSchedulerITCase` to include parallelism override tests

## Does this pull request potentially affect one of the following parts:

- Dependencies (does it add or upgrade a dependency): no
- The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
- The serializers: no
- The runtime per-record code paths (performance sensitive): no
- Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes (JobManager scheduler factories)
- The S3 file system connector: no

## Documentation

- Does this pull request introduce a new feature? no
- If yes, how is the feature documented? not applicable

barakbaron12 committed:
[FLINK-38770] Fix parallelism overrides ignored in Application Mode

This commit fixes a bug where pipeline.jobvertex-parallelism-overrides
configuration was completely ignored when running jobs in Application Mode
with Flink 2.0+. This broke the Flink Kubernetes Operator's autoscaler
functionality.

Root Cause:
- FLINK-36446 introduced direct StreamGraph submission in Application Mode
- Parallelism overrides were only applied in Dispatcher before JobGraph
  creation, but in Application Mode, StreamGraph→JobGraph conversion happens
  in scheduler factories, bypassing the Dispatcher's override logic

Solution:
- Created ParallelismOverrideUtil to centralize override application logic
- Modified all scheduler factories (DefaultScheduler, AdaptiveScheduler,
  AdaptiveBatchScheduler) to apply overrides after StreamGraph→JobGraph
  conversion
- Removed duplicate override logic from Dispatcher.java
- Ensured correct configuration precedence: job config > job master config

Testing:
- Added 7 unit tests for ParallelismOverrideUtil covering all scenarios
- Added 4 integration tests using MiniCluster to verify end-to-end behavior
  across different scheduler types
- All tests pass successfully

barakbaron12 committed:
[FLINK-38770] Apply parallelism overrides in AdaptiveBatchScheduler

This commit extends the fix for FLINK-38770 to AdaptiveBatchScheduler,
ensuring parallelism overrides work correctly in Application Mode for
all scheduler types.

Changes:
- Added ParallelismOverrideUtil.applyParallelismOverrides() call to
  AdaptiveBatchSchedulerFactory after StreamGraph to JobGraph conversion
- Refactored AdaptiveBatchSchedulerFactory to eliminate duplicate type
  checking and handle both JobGraph and StreamGraph in a single pass
- Fixed architecture violation by using getJobGraph(userCodeLoader)
  instead of getJobGraph() (which is marked @VisibleForTesting)
- Added test coverage for parallelism overrides in Application Mode:
  * ParallelismOverridesITCase: Tests with Default scheduler
  * AdaptiveSchedulerITCase: Tests with Adaptive scheduler
  * AdaptiveBatchSchedulerITCase: Tests with AdaptiveBatch scheduler
- All tests use RichMapFunction to capture and verify actual runtime
  parallelism, ensuring overrides are applied correctly
- Added @BeforeEach/@Before methods to reset test state between runs
- Removed inline comments from tests for cleaner code

All 22 tests pass successfully.

barakbaron12 commented:
Thanks for the suggestion @DongyuanPan!

I considered that approach but it doesn't quite work here. The main issue is that `StreamingJobGraphGenerator` only has access to the job-side configuration (via `StreamGraph.getJobConfiguration()`), but parallelism overrides can also come from the cluster-side config (`jobMasterConfiguration`). The override util merges both sources with the right precedence, and the generator simply doesn't have visibility into the cluster config — nor should it, since it's meant to be a pure structural converter.

Also, for `AdaptiveBatchScheduler`, the JobGraph can be built incrementally (vertex by vertex as upstream tasks finish), so applying overrides during generation would get messy.

The scheduler factories are really the first place where both the completed JobGraph and the cluster configuration exist together, which is why I went with applying overrides there.

barakbaron12 committed:
[FLINK-38770] Fix AdaptiveBatchScheduler: don't eagerly convert StreamGraph

The previous change broke the adaptive execution path by converting
StreamGraph to JobGraph before passing to createScheduler. This caused
AdaptiveExecutionHandlerFactory to always create NonAdaptiveExecutionHandler
instead of DefaultAdaptiveExecutionHandler, disabling adaptive parallelism.

Fix: only apply parallelism overrides for JobGraph inputs in createInstance,
and pass parallelism overrides through to DefaultAdaptiveExecutionHandler
for the adaptive StreamGraph path.

barakbaron12 commented:
Thanks for the follow-up @DongyuanPan, I think I understand now. If I'm reading you correctly, the idea is to merge the cluster-side overrides into the StreamGraph/JobGraph's own configuration inside `internalSubmitJob` (since the Dispatcher has access to both), so that all downstream code only needs to read from a single config source. Let me know if I'm misunderstanding.

I see the appeal, it would centralize the merge logic in one place. I do have a couple of concerns though:

1. Recovery path: `runRecoveredJob` calls `createJobMasterRunner` directly without going through `internalSubmitJob`. So if a job recovers after a failover, the cluster-side overrides wouldn't get merged into the plan's config. We'd either need to duplicate the merge logic in the recovery path (which defeats the single-place benefit), or accept that recovered jobs use whatever overrides were baked in at original submission time. With the current approach, the scheduler factories always merge fresh cluster config on every run, whether it's a new submission or a recovery.

2. Mutating the persisted plan: `persistAndRunJob` writes the execution plan to storage before running it. If we merge cluster overrides into the plan's config first, the persisted version includes those operational overrides. This means you can no longer tell from the persisted plan which overrides came from the job author vs. the cluster operator. Keeping them separate until the scheduler merges them preserves that distinction.

3. Adaptive batch complexity: As you noted, the `AdaptiveBatchScheduler` builds the JobGraph incrementally, vertices are created one-by-one as upstream tasks finish. The override check needs to happen per-vertex at decision time (in `getInitialParallelism()`), because that's when the scheduler decides between "use the override" vs "decide parallelism adaptively based on upstream data volume." This per-vertex logic stays the same regardless of where the config merge happens, so moving it to the Dispatcher doesn't simplify the adaptive path, which is really the trickiest part of this change.

4. The downstream logic stays the same either way: The scheduler factories still need to call `applyParallelismOverrides()` to set vertex parallelism, and `DefaultAdaptiveExecutionHandler.getInitialParallelism()` still needs the per-vertex override check for the adaptive path. The only difference is whether they read from one source or merge two, the application logic doesn't change.

Happy to discuss further if you see a way around these, or if I'm misunderstanding your suggestion!

voxeljorge commented:
Hi, we're very interested to see this land. We have backported this into Flink 2.0.1 for our own production environment and the fix seems to be working well. This has enabled us to stay on Flink 2.0.1 rather than downgrade after finding the autoscaler bug, although we'd like to upgrade further as soon as this actually lands.

DongyuanPan commented on flink-tests/src/test/java/org/apache/flink/test/scheduling/ParallelismOverridesITCase.java:
```diff
@@ -0,0 +1,318 @@
+     * parallelism overrides.
+     */
+    @Test
+    void testJobConfigurationTakesPrecedenceOverClusterConfiguration() throws Exception {
```
Unit tests can include a scenario where there is a conflict between the cluster and the job.

barakbaron12 replied:
Good catch — addressed in commit e1947768f57 by adding `testJobOverrideTakesPrecedenceOverClusterOverride`. The test configures a cluster-level override (parallelism 2) for a reserved vertex ID in `createConfiguration()`, then sets a job-level override (parallelism 4) for the same vertex, and asserts the job-level value wins. Thanks for the suggestion!

DongyuanPan reviewed (approved):
LGTM

barakbaron12 committed:
[FLINK-38770] Add test for cluster-vs-job override conflict resolution

Adds testJobOverrideTakesPrecedenceOverClusterOverride to verify that
when the same vertex has an override in both cluster-level and job-level
configurations, the job-level value wins.

davidradl reviewed (commented):
My AI says:
4 issues identified (1 medium, 3 low severity:
Medium: Missing error handling for invalid parallelism values in [ParallelismOverrideUtil.java:78](vscode-webview://1s6frf44ncago8fhff86hgrp390n8619rgsdbc7htn1tll7l1470/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/ParallelismOverrideUtil.java:78)
Low: Inconsistent placement in [AdaptiveBatchSchedulerFactory.java:131](vscode-webview://1s6frf44ncago8fhff86hgrp390n8619rgsdbc7htn1tll7l1470/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/AdaptiveBatchSchedulerFactory.java:131)
Low: Documentation could be enhanced in [ParallelismOverrideUtil.java:33](vscode-webview://1s6frf44ncago8fhff86hgrp390n8619rgsdbc7htn1tll7l1470/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/ParallelismOverrideUtil.java:33)
Low: Comment could reference FLINK-38770 in [DefaultSchedulerFactory.java:98](vscode-webview://1s6frf44ncago8fhff86hgrp390n8619rgsdbc7htn1tll7l1470/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultSchedulerFactory.java:98)

rmetzger commented:
> This PR fixes a bug where pipeline.jobvertex-parallelism-overrides configuration was completely ignored when running jobs in Application Mode with Flink 2.0+. This broke the Flink Kubernetes Operator's autoscaler functionality.

Why did no e2e or other test catch this?

barakbaron12 committed:
[FLINK-38770] Address review feedback: validation, docs, call-site consistency

- Validate parallelism override values with clear error messages (non-numeric
  and non-positive inputs now throw IllegalArgumentException with context
  instead of a generic NumberFormatException).
- Add applyParallelismOverridesIfApplicable(ExecutionPlan, Configuration)
  overload that no-ops for StreamGraph inputs, letting
  AdaptiveBatchSchedulerFactory call the utility at the same structural
  position as the other scheduler factories.
- Expand ParallelismOverrideUtil class javadoc to cover configuration
  precedence, call sites, and the adaptive-batch StreamGraph path.
- Add tests for the new validation and the ExecutionPlan overload.

barakbaron12 commented:
Quick update to both @davidradl and @rmetzger:

@davidradl - thanks for the review! I've addressed the feedback in commit e1947768f57:

1. **Medium - invalid parallelism values**: Added validation. Non-numeric values and non-positive integers now throw `IllegalArgumentException` with a clear message identifying the vertex and the invalid value. Added tests for both cases.

2. **Low - inconsistent placement in `AdaptiveBatchSchedulerFactory`**: Refactored to use a new overload `ParallelismOverrideUtil.applyParallelismOverridesIfApplicable(ExecutionPlan, ...)` that no-ops for `StreamGraph` inputs. This lets `AdaptiveBatchSchedulerFactory` call the utility at the same structural position as the other two scheduler factories, outside the type-check branch. That said, the new approach isn't perfect either, the overload hides the branching logic behind a method name, which is arguably still "inconsistent" in a different way. The `IfApplicable` suffix tries to hint that it may no-op, but readers still need to read the javadoc to understand why. If you have a better pattern in mind I'd really appreciate your thoughts!

3. **Low - documentation**: Expanded the class javadoc to cover what the feature does, the two config sources and their precedence, where the utility is called from, and the adaptive-batch `StreamGraph` path handled by `DefaultAdaptiveExecutionHandler`.

4. **Low - eference FLINK-38770 in the inline comment**: I'd gently push back on this one. JIRA refs don't seem to be common in inline comments in the codebase — they tend to live in commit messages, PR descriptions, and the public issue tracker. My concern is that references in source code can get stale when issues get moved/renamed. Happy to add it if you feel strongly.

@rmetzger - The honest answer is that there were no tests covering parallelism overrides in Application Mode before this PR. All existing tests submitted `JobGraph`s directly (Session Mode), which goes through the code path where overrides were applied in the Dispatcher. Application Mode submits a `StreamGraph` and the override call was never hit for that path, so the feature silently did nothing. The bug surfaced via the Flink Kubernetes Operator autoscaler, which uses this feature to dynamically adjust parallelism on a running application (as @voxeljorge mentioned above, they backported this fix to stay on 2.0.1 in production).

This PR closes that gap. We now have coverage for all three schedulers in Application Mode:
- `ParallelismOverridesITCase` - Default scheduler, including partial overrides and cluster-vs-job precedence
- `AdaptiveSchedulerITCase#testParallelismOverridesInApplicationMode`
- `AdaptiveBatchSchedulerITCase#testParallelismOverridesInApplicationMode`

I'm happy to add an end-to-end test if that would be useful, any suggestion on where it would fit best in `flink-end-to-end-tests`?

Dennis-Mircea reviewed (commented):
Thanks for bringing up this fix, @barakbaron12! The fix itself is correct. What needs clarification is the **impact statement**, which currently reads as "parallelism does not change" and "the operator autoscaler is broken". That is too broad: behavior depends on the deployment submode and the scaling path, and operator-side mechanics
already mask the bug in some configurations.

For context: [FLINK-38859](https://issues.apache.org/jira/browse/FLINK-38859) / [apache/flink-kubernetes-operator#1046](https://github.com/apache/flink-kubernetes-operator/pull/1046) shipped in operator **1.14.0** is an operator-side STANDALONE refinement that turns `max(overrides)` into the source of `parallelism.default` in addition to the existing TM-replica-driven path that already made STANDALONE rescale on every operator version. (cc @gyfora)

The below table summarizes the submode combinations that can be met by the Kubernetes Operator:

| Submode                                                       | Effect                                                                                                                                                                                                                                                                     | JIRA's "parallelism does not change" |
|---------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------|
| **STANDALONE** (any operator version)                         | The operator sets `KUBERNETES_TASKMANAGER_REPLICAS = numTaskManagers(parallelism)` and writes `parallelism.default` from `getParallelism()`, so parallelism does change end-to-end on rescale. Per-vertex granularity is lost (uniform `max`-style scaling).               | **Wrong**, parallelism does change.  |
| **NATIVE**, redeploy-based scaling                            | Overrides only land in `pipeline.jobvertex-parallelism-overrides`, dropped by `Dispatcher.internalSubmitJob` for `StreamGraph`. No operator-side fallback exists for NATIVE on any version.                                                                                | Correct.                             |
| **NATIVE**, in-place scaling (AdaptiveScheduler `/rescaling`) | Bypasses `Dispatcher.internalSubmitJob` entirely; per-vertex parallelism is applied via `JobMasterGateway`/`AdaptiveScheduler`.                                                                                                                                            | Bug does not apply.                  |
| **NATIVE** with `spec.taskManager.replicas` set               | `getParallelism()` returns `replicas * numTaskSlots` and writes it to `parallelism.default`. Bug masked **only if** something updates `replicas` on rescale (the autoscaler does NOT, `KubernetesScalingRealizer` only writes `pipeline.jobvertex-parallelism-overrides`). | Correct unless `replicas` changes.   |

The operator-version dimension only matters for **NATIVE**. STANDALONE has parallelism-driven scaling on every operator version the autoscaler supports.

## In-place scaling is unaffected

The whole analysis above only applies to the **redeploy-based** scaling path: the autoscaler writes overrides into `spec.flinkConfiguration`, the operator takes the job down and submits a fresh `StreamGraph` to a new JM, and that submit hits the broken `Dispatcher.internalSubmitJob`.

When **in-place scaling** is in use (autoscaler + Flink AdaptiveScheduler calling the JM's `/jobs/<id>/rescaling` REST endpoint), the existing JM applies the new per-vertex parallelism through `JobMasterGateway` / `AdaptiveScheduler` without ever re-submitting the job. There is no `StreamGraph` to `JobGraph` conversion and no second pass through `Dispatcher.internalSubmitJob`, so FLINK-38770 does not apply: per-vertex overrides are honored exactly. This is why NATIVE 2.x setups using in-place rescaling do not observe the bug despite having no operator-side workaround.

## Operator-side `getParallelism()` priority (latest version)

`FlinkConfigBuilder.getParallelism()` decides what gets written to
`parallelism.default`:

```java
// 1) TaskManager replicas-based (any mode)
if (spec.getTaskManager() != null && spec.getTaskManager().getReplicas() != null) {
    return spec.getTaskManager().getReplicas() * numTaskSlots;
}
// 2) STANDALONE override-based fallback, added by FLINK-38859 in operator 1.14.0+
if (KubernetesDeploymentMode.STANDALONE.equals(spec.getMode())) {
    Optional<Integer> max = getMaxParallelismFromOverrideConfig();
    if (max.isPresent() && max.get() > 0) return max.get();
}
// 3) Fall back to whatever the user declared
return spec.getJob().getParallelism();
```

In addition, `applyTaskManagerSpec` for STANDALONE Application Mode sets `KUBERNETES_TASKMANAGER_REPLICAS = numTaskManagers(parallelism)` whenever the user did not pin `taskManager.replicas` explicitly. That mechanism has always made STANDALONE rescale end-to-end, independent of FLINK-38859.

Implications:

- **STANDALONE** has working end-to-end scaling on every operator version the autoscaler supports, but only at uniform `max(overrides)` granularity. FLINK-38859 does not change that visible behavior; it only sources the number more directly from the override map.
- **NATIVE** has no such fallback at any operator version: parallelism is taken from `spec.job.parallelism`, which the autoscaler never updates. The redeploy path therefore drops overrides on the floor.
- **Branch (1)** masks the bug in either submode, but only if something updates `spec.taskManager.replicas` on rescale. `KubernetesScalingRealizer` does **not**, it only writes `pipeline.jobvertex-parallelism-overrides`. So branch (1) only "saves" you if a human or external tool bumps `replicas`.

## Why STANDALONE masks the bug at the StreamGraph level

The parallelism returned by `getParallelism()` is written to `parallelism.default`, so the Application Driver builds the StreamGraph with the new parallelism baked in **before** it reaches the Dispatcher. The broken dispatcher path is still exercised, but nothing depends on it anymore.

This works only when overrides are **uniform** (every vertex same value). On **non-uniform** scaling (e.g. `A:4, B:1`), `max = 4` is applied to every vertex and the light vertex (B) is silently over-provisioned to 4.

## Suggested impact wording for the JIRA / PR

> *In Flink 2.0+ Application Mode, `Dispatcher.internalSubmitJob` only applies `pipeline.jobvertex-parallelism-overrides` for `JobGraph` submissions, but Application Mode submits a `StreamGraph` (FLINK-36446). Effect on the operator autoscaler depends on submode and scaling path:*
>
> - *STANDALONE (any operator version): the operator computes TM replicas and `parallelism.default` from the autoscaler's overrides, so parallelism does change end-to-end. Per-vertex granularity is silently lost (works for uniform scaling, over-provisions on non-uniform). Operator >= 1.14.0 sources the value directly from `pipeline.jobvertex-parallelism-overrides` via FLINK-38859, but the visible "parallelism changes" behavior predates that.*
> - *NATIVE (redeploy-based scaling, any operator version): parallelism does not change, overrides are silently ignored. In-place scaling via AdaptiveScheduler `/rescaling` REST is unaffected.*

davidradl commented:
@XComp As per our slack. Are you ok with this ?

XComp commented:
Sorry for not getting back to this PR earlier. The notification slipped through, I guess.

> When in-place scaling is in use (autoscaler + Flink AdaptiveScheduler calling the JM's /jobs/<id>/rescaling REST endpoint), the existing JM applies the new per-vertex parallelism through JobMasterGateway / AdaptiveScheduler without ever re-submitting the job. 

The rescaling REST endpoint doesn't work. I guess, you're referring to the `job/<job-id>/resource-requirements` endpoint that was introduced with [FLIP-291](https://cwiki.apache.org/confluence/display/FLINK/FLIP-291%3A+Externalized+Declarative+Resource+Management)?

I will look into the PR as well.

pavlo-holinko commented:
Any chance to get this merged anytime soon? Happy to help if needed.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/DefaultAdaptiveExecutionHandler.java:
```diff
@@ -69,6 +78,14 @@ public DefaultAdaptiveExecutionHandler(
         this.streamGraphOptimizer.initializeStrategies(
                 adaptiveGraphManager.getStreamGraphContext());
+
+        // Merge overrides: master config first, job config wins (same precedence as
```
We're skipping the validation of the overrides here which would happen in `ParallelismOverrideUtil#parseParallelism§` I guess that should be fixed before merging, hm?

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/AdaptiveBatchSchedulerFactory.java:
```diff
@@ -122,7 +123,10 @@ public SchedulerNG createInstance(
-        ExecutionConfig executionConfig;
+        final ExecutionConfig executionConfig;
+
+        ParallelismOverrideUtil.applyParallelismOverridesIfApplicable(
+                executionPlan, jobMasterConfiguration);
```
Wouldn't it make sense to move this also into `AdaptiveExecutionHandlerFactory`. I mean we would still have the split of responsibilities between `AdaptiveExecutionHandlerFactory` and `DefaultAdaptiveExecutionHandler`. But we would at least make it a bit more consistent. WDYT?

We could delete `ParallelismOverrideUtil#applyParallelismOverridesIfApplicable` then.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/AdaptiveExecutionHandlerFactory.java:
```diff
@@ -54,18 +58,25 @@ public static AdaptiveExecutionHandler create(
+                        userClassLoader,
+                        (StreamGraph) executionPlan,
+                        serializationExecutor,
+                        jobMasterConfiguration);
```
Instead of passing in `jobMasterConfiguration`, we could also extract the per-vertex parallelism overrides here and pass them into `DefaultAdaptiveExecutionHandler`. That way (with the [other suggestion above](https://github.com/apache/flink/pull/27334/changes#r3566978850) considered), we could move the override handling entirely into `AdaptiveExecutionHandlerFactory`.

XComp commented on flink-tests/src/test/java/org/apache/flink/test/scheduling/AdaptiveBatchSchedulerITCase.java:
```diff
@@ -404,4 +413,107 @@ public boolean onOperatorsFinished(
     }
+
+    /**
+     * Tests that parallelism overrides work correctly with the AdaptiveBatch scheduler. This
```
```suggestion
     * Tests that parallelism overrides work correctly with the {@code AdaptiveBatchScheduler}. This
```
nit

XComp commented on flink-tests/src/test/java/org/apache/flink/test/scheduling/AdaptiveBatchSchedulerITCase.java:
```diff
@@ -404,4 +413,107 @@ public boolean onOperatorsFinished(
+        jobGraph.getJobConfiguration().set(PipelineOptions.PARALLELISM_OVERRIDES, overrides);
+
+        // Submit and run the job - it will use the overridden parallelism (2 slots needed)
+        JobGraphRunningUtil.execute(jobGraph, configuration, 1, 2);
```
We have to find a better assert here: The test also succeeds if the override is not applied. I guess that's due to the fact that the job will just executed in multiple batches.

XComp reviewed (commented):
Sorry for not getting back to this PR earlier. I finally found time to go through it. I have a few comments around test coverage and cosmetic changes. Overall the code change is reasonable. I'm happy to merge it after the open questions are addressed/discussed.

rmetzger commented:
@barakbaron12 are you still working on this?
